Introduce memory stacks for free

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107171
* __->__ #106758

Previously when we recorded a free action in a memory trace, we would provide
the stack for when the block was allocated. This is faster because we do not
have to record stacks for free, which would otherwise double the number of stacks
collected. However, sometimes knowing the location of a free is useful for
figuring out why a tensor was live. So this PR adds this behavior. If
performance ends up being a concern the old behavior is possible by passing
"alloc" to the context argument rather than "all".

Also refactors some of glue logic to be consistent across C++ and Python and
routes the Python API through the C++ version.

zdevito committed:
Introduce memory stacks for free

Previously when we recorded a free action in a memory trace, we would provide
the stack for when the block was allocated. This is faster because we do not
have to record stacks for free, which would otherwise double the number of stacks
collected. However, sometimes knowing the location of a free is useful for
figuring out why a tensor was live. So this PR adds this behavior. If
performance ends up being a concern the old behavior is possible by passing
"alloc" to the context argument rather than "all".

Also refactors some of glue logic to be consistent across C++ and Python and
routes the Python API through the C++ version.

[ghstack-poisoned]

zdevito committed:
Update on "Introduce memory stacks for free"


Previously when we recorded a free action in a memory trace, we would provide
the stack for when the block was allocated. This is faster because we do not
have to record stacks for free, which would otherwise double the number of stacks
collected. However, sometimes knowing the location of a free is useful for
figuring out why a tensor was live. So this PR adds this behavior. If
performance ends up being a concern the old behavior is possible by passing
"alloc" to the context argument rather than "all".

Also refactors some of glue logic to be consistent across C++ and Python and
routes the Python API through the C++ version.

[ghstack-poisoned]

albanD commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -105,17 +105,66 @@ void _record_memory_history(
+  max_entries = (enabled && *enabled == "all") ? max_entries : 1;
+  auto when = c10::cuda::CUDACachingAllocator::RecordContext::NEVER;
+  if (context) {
+    if (context == "all") {
```
nit: you can also just bind the enum if you want to avoid this/reduce string on the API: https://github.com/pytorch/pytorch/blob/ad22f0ffb456fc3f967ad32e09376f7c9cf94a56/torch/csrc/Module.cpp#L1496-L1519

zdevito replied:
I want to avoid including the allocator header file in memory_snapshot.h because it slowdown recompilation of the allocator. I also think it makes sense that the C++ and Python API for recording history has the same types.

albanD commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -181,6 +181,13 @@ struct CheckpointDelta {
 
+enum struct RecordContext {
+  NEVER = 0,
+  STATE = 1, // only keep stacks only for active allocations
```
```suggestion
  STATE = 1, // only keep stacks for active allocations
```

albanD reviewed (commented):
Sounds good.
It's not so easy by looking at the diff to tell what happens for delayed free? The stack will point at the original free call? Could we track both? :D

zdevito committed:
Update on "Introduce memory stacks for free"


Previously when we recorded a free action in a memory trace, we would provide
the stack for when the block was allocated. This is faster because we do not
have to record stacks for free, which would otherwise double the number of stacks
collected. However, sometimes knowing the location of a free is useful for
figuring out why a tensor was live. So this PR adds this behavior. If
performance ends up being a concern the old behavior is possible by passing
"alloc" to the context argument rather than "all".

Also refactors some of glue logic to be consistent across C++ and Python and
routes the Python API through the C++ version.

[ghstack-poisoned]

zdevito committed:
Update on "Introduce memory stacks for free"


Previously when we recorded a free action in a memory trace, we would provide
the stack for when the block was allocated. This is faster because we do not
have to record stacks for free, which would otherwise double the number of stacks
collected. However, sometimes knowing the location of a free is useful for
figuring out why a tensor was live. So this PR adds this behavior. If
performance ends up being a concern the old behavior is possible by passing
"alloc" to the context argument rather than "all".

Also refactors some of glue logic to be consistent across C++ and Python and
routes the Python API through the C++ version.

[ghstack-poisoned]

albanD reviewed (approved):
Sounds perfect! Thanks a lot!

zdevito committed:
Update on "Introduce memory stacks for free"


Previously when we recorded a free action in a memory trace, we would provide
the stack for when the block was allocated. This is faster because we do not
have to record stacks for free, which would otherwise double the number of stacks
collected. However, sometimes knowing the location of a free is useful for
figuring out why a tensor was live. So this PR adds this behavior. If
performance ends up being a concern the old behavior is possible by passing
"alloc" to the context argument rather than "all".

Also refactors some of glue logic to be consistent across C++ and Python and
routes the Python API through the C++ version.

[ghstack-poisoned]

ngimel commented:
+1 for the free stacks!

zdevito commented:
@pytorchbot merge
